openvdb_ax/openvdb_ax/ast/Tokens.h

```diff
@@ -0,0 +1,364 @@
+    }
+    else if (type == "bool")   return BOOL;
+    else if (type == "char")   return CHAR;
+    else if (type == "short")  return SHORT;
+    else if (type == "int")    return INT;
+    else if (type == "long")   return LONG;
+    else if (type == "float")  return FLOAT;
+    else if (type == "double") return DOUBLE;
+    else if (type == "string") return STRING;
+
+    // also handle vdb types that have different type strings to our tokens
+    // @todo  These should probably be separated out. The executables currently
+    //   use this function to guarantee conversion
+    if (type[0] == 'v') {
+        if (type == "vec2s") return VEC2F;
+        if (type == "vec3s") return VEC3F;
+        if (type == "vec4s") return VEC4F;
+    }
+    else if (type[0] == 'm') {
+        if (type == "mat3s") return MAT3F;
+        if (type == "mat4s") return MAT4F;
+    }
+    else if (type == "quats") return QUATF;
+    else if (type[0] == 'i') {
+        if (type == "int16") return SHORT;
+        if (type == "int32") return INT;
+        if (type == "int64") return LONG;
+    }
+
+    return UNKNOWN;
+}
+
+inline std::string typeStringFromToken(const CoreType type)
+{
+    switch (type) {
+        case BOOL    : return "bool";
+        case CHAR    : return "char";
+        case SHORT   : return "short";
+        case INT     : return "int";
+        case LONG    : return "long";
+        case FLOAT   : return "float";
+        case DOUBLE  : return "double";
+        case VEC2I   : return "vec2i";
+        case VEC2F   : return "vec2f";
+        case VEC2D   : return "vec2d";
+        case VEC3I   : return "vec3i";
+        case VEC3F   : return "vec3f";
+        case VEC3D   : return "vec3d";
+        case VEC4I   : return "vec4i";
+        case VEC4F   : return "vec4f";
+        case VEC4D   : return "vec4d";
+        case MAT3F   : return "mat3f";
+        case MAT3D   : return "mat3d";
+        case MAT4F   : return "mat4f";
+        case MAT4D   : return "mat4d";
+        case QUATF   : return "quatf";
+        case QUATD   : return "quatd";
+        case STRING  : return "string";
+        case UNKNOWN :
+        default      :
+            return "unknown";
+    }
+}
+
+template<typename T> CoreType tokenFromType()                    { return UNKNOWN; }
+template<> inline CoreType tokenFromType<bool>()                 { return BOOL; }
+template<> inline CoreType tokenFromType<char>()                 { return CHAR; }
+template<> inline CoreType tokenFromType<int16_t>()              { return SHORT; }
+template<> inline CoreType tokenFromType<int32_t>()              { return INT; }
+template<> inline CoreType tokenFromType<int64_t>()              { return LONG; }
+template<> inline CoreType tokenFromType<float>()                { return FLOAT; }
+template<> inline CoreType tokenFromType<double>()               { return DOUBLE; }
+template<> inline CoreType tokenFromType<math::Vec2<int32_t>>()  { return VEC2I; }
```

Comment:
Off the back of this comment we decided to remove `short`/`int16` local variable support and allow `int` to be aliased to 32 or 64bit. There remains the question of what to do with integer literals as it would be nice to remove the `l` suffix for longer types. There's two options in my mind:
1) all literals are 64 bit regardless of your `int` type selection. This is slightly strange as literal arithmetic may be done at a higher precision. Additionally, because other container types like vec3i are 32bit, you get implicit warnings with things like `vec3i a = {1,2,3};` or `int a = 1;` unless we explicitly ignore assignments from int literals. 
2) literals are automatically stored as 64 bit if the value is larger than `limit<int32_t>::max()`. This is nicer, but has the strange side effect of producing different results if, say, an evaluation of two literals produces a 64bit value. i.e. `1 + 2147483648` vs `1 + 2147483649`. The latter will return a int64 value, the prior will wrap and return a 32bit value.
Otherwise, if literals are tied to the type alias of `int` (e.g. 32bit) then we'd have to keep a suffix to distinguish 64bit values

Reply:
I'm most likely not going to add this integer precision switch for this PR. I have removed `int16` types but haven't been able to come up with an implementation that I'm happy with yet w.r.t aliasing int to either 32 or 64 bit. This would be much simpler if a) AX wasn't bound to closely to VDB types and b) floats were treated in a similar way. We can add this later at no change to the frontend and I'll be able to address all three of these issues together which will make for a much better implementation

Reply:
Sounds good!